### PR TITLE
utils: Change disjoint() to return common elements

### PR DESCRIPTION
**Pull request by sschuberth**
This helps to identify the common elements.

Signed-off-by: Sebastian Schuberth <sebastian.schuberth@here.com>

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/1399)

### CODE REVIEW

File: utils/src/main/kotlin/Utils.kt [hunk 1]
```diff
@@ -59,18 +58,21 @@ const val PARAMETER_ORDER_LOGGING = 2
 const val PARAMETER_ORDER_HELP = 100
 
 /**
- * Check whether the specified two or more collections have no elemets in common.
+ * Return the set of elements the specified two or more collections have in common, or an empty set if all collections
+ * are disjoint.
  */
-fun disjoint(c1: Collection<*>, c2: Collection<*>, vararg cN: Collection<*>): Boolean {
+fun <T> disjoint(c1: Collection<T>, c2: Collection<T>, vararg cN: Collection<T>): Set<T> {
```
Comment (fviernau) [inline utils/src/main/kotlin/Utils.kt]: It now computes the intersection, how about renaming it to "intersect" ?

File: utils/src/main/kotlin/Utils.kt [hunk 2]
```diff
@@ -59,18 +58,21 @@ const val PARAMETER_ORDER_LOGGING = 2
...
     val c = listOf(c1, c2, *cN)
 
+    val commonElememts = mutableSetOf<T>()
```
Comment (fviernau) [inline utils/src/main/kotlin/Utils.kt]: I'm not sure about the sizes of the collection usually passed to this function,
if sizes are not small consider making it more efficient, like:
```
val list = listOf(c2, *cN)
var commonElements = c[a]

for (i in list.indices) {
    commonElements = commonElements.intersect(list[i])
}
```

File: utils/src/test/kotlin/UtilsTest.kt [hunk 3]
```diff
@@ -29,11 +29,11 @@ import java.nio.file.Paths
 class UtilsTest : WordSpec({
     "disjoint" should {
         "return true for collections that have no common elememts" {
```
Comment (fviernau) [inline utils/src/test/kotlin/UtilsTest.kt]: Return type is not boolean anymore, please adjust the test name accordingly.

File: utils/src/test/kotlin/UtilsTest.kt [hunk 4]
```diff
@@ -29,11 +29,11 @@ import java.nio.file.Paths
...
-            disjoint(setOf(1), setOf(2), setOf(3)) shouldBe true
+            disjoint(setOf(1), setOf(2), setOf(3)) shouldBe emptySet()
         }
 
         "return false for collections that have common elememts" {
-            disjoint(setOf(1, 2), setOf(2, 3), setOf(3, 4)) shouldBe false
+            disjoint(setOf(1, 2), setOf(2, 3), setOf(3, 4)) shouldBe setOf(2, 3)
```
Comment (fviernau) [inline utils/src/test/kotlin/UtilsTest.kt]: According to the function documention - shouldn't this return an empty set ?

File: utils/src/main/kotlin/Utils.kt (hunk 1 above)
Comment (sschuberth) [reply utils/src/main/kotlin/Utils.kt]: I made the same mistake in my first iteration of this PR 😉 It does *not* compute the intersection of all collections, but it checks whether any two collections have a non-empty intersection. That's a difference! In that sense, the goal of the function still is to check whether all collections are disjoint.

File: utils/src/main/kotlin/Utils.kt (hunk 2 above)
Comment (sschuberth) [reply utils/src/main/kotlin/Utils.kt]: See above, this computes something different.

File: utils/src/test/kotlin/UtilsTest.kt (hunk 4 above)
Comment (sschuberth) [reply utils/src/test/kotlin/UtilsTest.kt]: No, but I had to fix the test descriptions anyway, done, thanks.

File: utils/src/test/kotlin/UtilsTest.kt (hunk 3 above)
Comment (sschuberth) [reply utils/src/test/kotlin/UtilsTest.kt]: Done.

File: utils/src/main/kotlin/Utils.kt (hunk 1 above)
Comment (fviernau) [reply utils/src/main/kotlin/Utils.kt]: Oh alright, that's a different story then.

The function documentation in combination with the function name to me seems still misleading, maybe sth like the following makes it clear?

comment: "Return the set of elements which are contained in at least two of the given collections."
name: "allPairsDiscoint", "unitePairwiseIntersections" hm, finding a good name is quite tricky...

Comment (sschuberth) [reply utils/src/main/kotlin/Utils.kt]: I've improved the docs like you suggest, but I've kept the function name as it's still the same function as before, and I cannot think of a better name right now either 😉

Comment (fviernau) [reply utils/src/main/kotlin/Utils.kt]: Thanks!

**Comment by sschuberth** [comment]
AppVeyor seems to have some issues as it does not complete. Still merging as Travis completed successfully.

**Pull request merged**